### PR TITLE
Add 'The Developers Certificate of Origin' short article (#523)

### LINKED ISSUE #523: contributor license agreement policies and alternatives
I participate in some projects that use an ANL-approved CLA agreement based on the Apache Software Foundation's policies (https://www.apache.org/licenses/contributor-agreements.html).  These agreements provide assurance that contributors will not assert IP claims on their contributions.  This is meant to protect not just the maintainers, but also third party users and redistributors.

Conceptually that is fine, but there are two practical problems: a) these agreements are nearly impossible to get signed at some organizations and b) they greatly inhibit casual/agile contributions from users.

I would like to find a less-cumbersome alternative.  One likely option is the DCO https://developercertificate.org/.  Gitlab some years back switched to this model from a CLA and wrote up their analysis at https://docs.google.com/document/d/1zpjDzL7yhGBZz3_7jCjWLfRQ1Jryg1mlIVmG8y6B1_Q/edit.  From a process point of view, a DCO can be acknowledged by signing off on a git commit. It does not require engaging the legal department of the contributor's home organization.

I don't know if DCOs are used elsewhere in DOE projects, or if there are other alternatives to consider.  My objective is simply to find an accepted alternative that's a more practical middle ground between a CLA and nothing.

Note: this issue is separate from the choice of what software license to apply to the code itself (at least in the cases that I am familiar with).  I'm perfectly happy with standard 3 clause BSD licenses for that purpose.

### PR DESCRIPTION
**Pull request by bartlettroscoe**
# Description

EB Member: @bernhold 

Resolves #523

Turned out to be a short article instead of a CC article for #523.

## PR checklist for files displayed on bssw.io site

* [x] ***[Author]*** `@mention` the BSSw.io editorial board member `@<eb-member-id>` in **Description** above assigned to shepherd your PR.
* [x] ***[Author]*** Add the `<issue-id>` in the **Description** above for the associated GitHub Issue.
* [x] ***[Author]*** Ensure `wikize_refs.py -i <base>.md` is run and commit (if using [wikize_refs.py]).
* [x] ***[Author]*** Inspect the content in the `*.md` file(s) as rendered in GitHub for this PR.
* [x] ***[EB Mem]*** Assign this PR to the EB member `<eb-member-id>`.
* [x] ***[EB Mem]*** Assign this PR to the author of the PR `<pr-author-id>`.
* [x] ***[EB Mem]*** Add one or more Reviewers.
* [x] ***[EB Mem]*** Add label `content: <content-type>` for the type of contribution.
* [x] ***[EB Mem]*** Add to Project `Content Development` (see [Content Development]).
* [x] ***[EB Mem]*** Add [meta-data] to the `*.md` file(s) (set `Publish: yes`).
* [x] ***[EB Mem]*** Add label `preview` (so PR branch will be merged to 'preview' branch and watch for possible merge failures).
* [x] ***[EB Mem]*** Rebuild [preview] site and confirm new content is there, renders correctly and is returned in searches.
* [x] ***[Author]*** Make any final changes to the PR based on feedback.
* ~~[ ] ***[Author]*** Ensure `wikize_refs.py -i <base>.md` is run and commit (if using [wikize_refs.py]).~~
* [x] ***[EB Mem]*** Rebuild [preview] site and re-confirm content looks correct.
* [x] ***[EB Mem]*** Ensure at least one reviewer signs off on the final changes.
* [x] ***[EB Mem]*** Move the PR to "Ready to Publish" in [Content Development].
* [ ] ***[EB Mem]*** Leave comment and @mention `@betterscientificsoftware/bssw-maint` (**BSSw Maint**) asking to carry out final publication steps.
* [ ] ***[BSSw Maint]*** Verify that all needed files are present in the PR (article, images, updates to Site/Homepage.md carousel and/or Site/Announcements/Announcements.md as appropriate).
* [ ] ***[BSSw Maint]*** Merge PR. (Should automatically move to "Done" in [Content Development].)
* [ ] ***[BSSw Maint]*** Verify new contribution shows up on [bssw.io] as expected.

NOTE:
* Checklist items prefixed with ***[Author]*** are expected to be performed by the author of the PR or can be performed by the author.
* Checklist items prefixed with ***[EB Mem]*** must be performed by a [BSSw.io Editorial Board (EB) Member](https://betterscientificsoftware.github.io/bssw.io/bssw_members.html).
* Checklist items prefixed with ***[BSSw Maint]*** must be performed by someone in the `@betterscientificsoftware/bssw-maint` team (hint: type `@`, `b`, `s`, `s`, `w`, `-`, `m`  to auto-complete to `@betterscientificsoftware/bssw-maint`).

[preview]: https://preview.bssw.io
[bssw.io]: https://bssw.io
[Content Development]: https://github.com/betterscientificsoftware/bssw.io/projects/3
[BSSw Internal]: https://github.com/betterscientificsoftware/bssw.io/projects/2
[meta-data]: https://betterscientificsoftware.github.io/bssw.io/bssw_styling_common.html#metadata-section
[wikize_refs.py]: https://github.com/betterscientificsoftware/bssw.io/blob/master/utils/README.md#wikize_refspy

### CODE REVIEW

**Comment by bernhold** [comment]
Hi Ross, thanks for this.  I'm going to provide some deeper feedback shortly.  But to let you know, due to some personal (or personnel, depending on how you look at it) issues, we had to move up the publication of the digest and it has actually just appeared in my mailbox as I was typing this.  So I expect we'll highlight your article in January.

**Comment by bartlettroscoe** [comment]
> Hi Ross, thanks for this. I'm going to provide some deeper feedback shortly.

@bernhold, thanks!  I think all of my edits are done and I think I should have the display issues on bssw.io worked out now.

> we had to move up the publication of the digest and it has actually just appeared in my mailbox as I was typing this. So I expect we'll highlight your article in January.

Understood.  Glad to see there were two other articles that were able to fit the bill for Dec.

**Comment by bartlettroscoe** [comment]
FYI: I think I have the display issues worked out on preview.bssw.io shown at:

* https://preview.bssw.io/items/the-developer-certificate-of-origin

for which I had to create several experimental commits shown [here](https://github.com/betterscientificsoftware/bssw.io/pull/1156/commits) with lots of rebuilds.

It turns out that the bssw.io site generator unconditionally strips text inside of angle brackets `<...>` even if those chars are in a verbatim field.  So in the case where I should be showing `<` and `>` I had to uses `{` and `}` which is not technically correct but I don't know what choice you have.

This non-standard buggy site generator for bssw.io should be addressed at some point.  It can really significantly increase overhead of getting an article published to bssw.io.

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (bernhold) [inline Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md:77]: Here I'd like to add

### Disclaimer

This is not legal advice. Consult your own lawyer before any action that may have legal consequences.  The article is based on U.S. copyright law; other countries may differ.

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (bernhold) [inline Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md:63]: Can we remove "{Met URL}" in line 63?  I don't think it is relevant here. We should just end the quotation at the end of the sentence.

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (bernhold) [inline Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md:12]: I propose changing the first paragraph to read as follows:

When a person writes a piece of software, by default, they own the copyrights in that software by default, unless it is a "work for hire", in which case their employer owns it, or those rights are coveyed to some other legal entity by a written agreement.
And even if that person provides that source code with the implied intent to make it part of some existing software project or otherwise give it away (i.e. place it in the public domain), legal experts may differ as to whether that by itself is enough to allow the receiving project to modify and distribute the contribution along with the rest of the project.  For some experts, contributions are implictly offered under the same licensing terms as the code base to which the contribution is being made.  Others prefer to have the owner of the copyright in new contributions 
explicitly state that they are providing the software (or a modification to an existing piece of software) to a project in accordance to some specific software license associated with the project<sup>[0]</sup>.
It is the process by which a person provides an (updated) contribution to an existing project that has an existing software license that is the focus of this article.

**Review by bernhold — CHANGES_REQUESTED** [review]
Hi Ross, I've suggested some changes.  Sorry if this isn't the best way to do it.

I'm trying to capture a bit of discussions I've had with our IP counsel about these matters.  His preference is that go with the simplicity of contributions being implicitly made under the same terms as the underlying code base.  He's okay with us (as individuals) signing off on DCOs because that's not a legal agreement per se.  Whereas CLAs are of course legal agreements and individuals are not empowered to sign them.  And that's where the problems with CLAs *begin*, as I think you know.

**Comment by bartlettroscoe** [comment]
> Whereas CLAs are of course legal agreements and individuals are not empowered to sign them. 

@bernhold, right, but if you are working a private person on an open source project on you own time apart from your company, then I think you could sign a CLA for those efforts.  (But I guess it depends on how your employment agreement is written.)

> And that's where the problems with CLAs begin, as I think you know.

Right, the major problem with CLA are legal issues while the major problems with the DCO are the technical mechanics of signing the commits and validated that the people signing the commits are who they say they are. 

I will post updated commits with all of your suggestions.

Thanks!

**Comment by bernhold** [comment]
> > Whereas CLAs are of course legal agreements and individuals are not empowered to sign them.
> 
> @bernhold, right, but if you are working a private person on an open source project on you own time apart from your company, then I think you could sign a CLA for those efforts. (But I guess it depends on how your employment agreement is written.)

ORNL owns *everything* I do that can be considered in any way related to my paid work. And it is not well defined just how far that extends.  For clarity, I'd probably want to request official approval for an Outside Activity.

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (bartlettroscoe) [reply Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md:83]: Should be addressed in new commit badd685f8f5bf9f11a7a345e495083b25ecd127e

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (bartlettroscoe) [reply Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md]: Should be addressed in new commit badd685f8f5bf9f11a7a345e495083b25ecd127e

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (bartlettroscoe) [reply Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md]: Should be addressed in new commit badd685f8f5bf9f11a7a345e495083b25ecd127e

**Comment by bartlettroscoe** [comment]
@bernhold, thanks for the suggestions.  I have incorporated all of them and rebuilt the preview site with the updated article at:

* https://preview.bssw.io/items/the-developer-certificate-of-origin

**Comment by bernhold** [comment]
@bartlettroscoe fyi, I dug deeper into the rendering issue you encountered.  See #1160 for a set of tests and comparison of how they are rendered with various tools.  BSSw uses https://github.com/vmg/redcarpet by the way.

I've started a discussion with Clara at Parallactic to see if we can get RedCarpet to be closer to the others without other adverse consequences (I suspect their :escape_html setting might be the key, but it sounds pretty general and I don't know if other things we might *not* want might come from turning it on).  

The tests show how to get the results you want in your code block either way.  But let's give it a little time to see if Clara can bring us closer to the behavior we expect before we change your markup.

**Comment by bartlettroscoe** [comment]
> The tests show how to get the results you want in your code block either way. But let's give it a little time to see if Clara can bring us closer to the behavior we expect before we change your markup.

@bernhold, it is good to know that it can be fixed.  I think what we want is to have the bssw.io site generator to be a close as possible to the way that GitHub renders Markdown.  We should likely spend some time identifying all of areas where there are differences and see if we can make them behave the same.  I think this is a really important issue.  It should be rare that you find a problem when you view the article on preview.bssw.io if it looks good as rendered on GitHub.  That is currently not the case.

Also, I think this is PR ready for final approval.

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (bernhold) [inline Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md:20]: I suggest ^allowing^facilitating. You can relicense either way.  But if your CLA includes a relicensing clause, it will probably be a *lot* easier.

**Review by bernhold — APPROVED** [review]
Just one more small change I've suggested below. 

Also, per a comment in the main conversation, we now know how to get the \<\> markup you want in the code block.  But let's hold on those changes to see whether we can get the rendered changed instead of having to accommodate it here.

Nice article.  Thanks!

**Comment by bernhold** [comment]
> I think what we want is to have the bssw.io site generator to be a close as possible to the way that GitHub renders Markdown. We should likely spend some time identifying all of areas where there are differences and see if we can make them behave the same. I think this is a really important issue.

I agree, this is the principle of least surprise.  I've made the point to Clara. We'll see what she says.

**Comment by bartlettroscoe** [comment]
> Also, per a comment in the main conversation, we now know how to get the `<>` markup you want in the code block. But let's hold on those changes to see whether we can get the rendered changed instead of having to accommodate it here.

@bernhold, okay, let's wait to merge this PR to see if they can fix the rendering of `< >`.  Then I can fix article to make it technically correct.

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (bartlettroscoe) [reply Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md]: Thanks!  Addressed in new commit ad48d7ea362084201f08e8da371dde0595dd11bf.

**Comment by bernhold** [comment]
@bartlettroscoe Clara has a fix for the <> issue.  I've confirmed it on the staging site and asked her to push it to production.  I'll try to let you know when that happens.  You should be able to independently verify through the test case I created in #1160 which is on preview.

**Comment by bernhold** [comment]
@bartlettroscoe this is now in production.  I've updated #1160 to show the new results -- everything is consistent with GH and VS Code's renderings.  You should be able to go back to your preferred markup in this article.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe this is now in production. I've updated #1160 to show the new results -- everything is consistent with GH and VS Code's renderings. You should be able to go back to your preferred markup in this article.

@bernhold, thanks for taking care of this.  I pushed the commit ??? bring back `< ... >` and rebuilt the preview site and can confirm the article now displays `< ... > ` correctly at:

* https://preview.bssw.io/items/the-developer-certificate-of-origin

showing:

image

and:

image

I think this is ready to publish.

**Comment by bartlettroscoe** [comment]
@betterscientificsoftware/bssw-maint,

This is ready to publish.   Can someone please carry out the final publication steps?

**Comment by rinkug** [comment]
@karbarz : could you give this a quick read-through. It would be great to publish it this week.

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (bartlettroscoe) [inline Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md:59]: I am not sure the comma  at "... including the Linux Kernel[8] and more recently, the source code for GitLab ..." makes sense.  If there is going to be a comma there, then there should also be a comma before the "and" as well, no?  So this would be:

"Some major projects are using the DCO with Git commit sign-offs for handling contributions, including the Linux Kernel<sup>[8]</sup>, and more recently, the source code for GitLab itself <sup>[3],[4]</sup>."

?

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (bartlettroscoe) [inline Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md:68]: Does a comma at "... for many developers and therefore, more limited usages ..." make sense?  I think, if you add a comma after "therefore" you need to add it before as well like:

"... for many developers and, therefore, more limited usages ..."

See:

* https://english.stackexchange.com/questions/246304/proper-use-of-comma-before-and-after-therefore/246385#:~:text=In%20your%20sample%20sentence%2C%20therefore,I%20need%20to%20respect%20him.

But I think it would be correct to leave the commas out around "therefore".

Other opinions?

**Review by bartlettroscoe — COMMENTED** [review]
@karbarz and @rinkug ,

I agree with almost all of these changes.  Mostly these are breaking up (possibly) run-on sentences and adding some commas.  But there are a few commas that are added that are noted below that might need some tweaking.

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (karbarz) [reply Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md:59]: Good catch here! I believe that the second comma would actually go after the "and", as we're breaking up the interjection of "more recently". So, it would look like:

"Some major projects are using the DCO with Git commit sign-offs for handling contributions, including the Linux Kernel[8] and **, more recently,** the source code for GitLab itself [3],[4]."

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (karbarz) [reply Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md:68]: I agree with you here. I believe I misread this portion and my brain automatically went into "comma" mode, but I dont' think it's necessary in this instance.

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (bartlettroscoe) [reply Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md]: Thanks! Should be resolved in my last commit 9ec2820627836ebe79c37b588db6556142b06f83.

File: Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md
Comment (bartlettroscoe) [reply Articles/ShortArticles/TheDeveloperCertificateOfOrigin.md]: Thanks! Should be resolved in my last commit 9ec2820627836ebe79c37b588db6556142b06f83.

**Comment by bartlettroscoe** [comment]
@rinkug and @bernhold, with the to commits from @karbarz and me, this should be ready to merge and publish.

**Pull request merged**
